### ISSUE TITLE
Moses can not be installed with OXLM

### ISSUE THREAD

**Issue by doried-a-a** [pauldb89/OxLM#3]
Hello
I'm trying to build moses with oxlm, I've been trying for many weeks with no results.
When I try to build oxlm following the instructions described on this repo, it builds successfully. All I have to do is to set $BOOST_ROOT= path to boost installation on my PC.
When I try to build moses, using the default configurations(static linking)..  OXLM can not find any boost library..  and everything fails
But when building moses with shared linking, specifically, with the following statement:

-----------------------------
./bjam link=shared --shared -a -j24 --with-oxlm=/home/doried/tran/oxlm/  | tee out.txt
-----------------------------
I get the following errors...  

Tip: install tcmalloc for faster threading.  See BUILD-INSTRUCTIONS.txt for more information.
NOT BUILDING MOSES SERVER!
Performing configuration checks

    - Shared Boost             : yes
    - Static Boost             : yes
error: No best alternative for ./boost_serialization
    next alternative: required properties: <link>shared
        matched
    next alternative: required properties: <link>shared
        matched
error: No best alternative for ./boost_serialization
    next alternative: required properties: <link>shared
        matched
    next alternative: required properties: <link>shared
        matched
error: No best alternative for ./boost_serialization
    next alternative: required properties: <link>shared
        matched
    next alternative: required properties: <link>shared
        matched
error: No best alternative for ./boost_serialization
    next alternative: required properties: <link>shared
        matched
    next alternative: required properties: <link>shared
        matched
error: No best alternative for ./boost_serialization
    next alternative: required properties: <link>shared
        matched
    next alternative: required properties: <link>shared
        matched
error: No best alternative for ./boost_serialization
    next alternative: required properties: <link>shared
        matched
    next alternative: required properties: <link>shared
        matched
error: No best alternative for ./boost_serialization
    next alternative: required properties: <link>shared
        matched
    next alternative: required properties: <link>shared
        matched
......
......
......
**HERE every thing seems to work normally**
......
......

**and finally there is something like:**

gcc.link moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/BackwardTest
/usr/bin/ld: moses/LM/bin/gcc-4.8/release/threading-multi/OxLM.o: undefined reference to symbol '_ZN5boost7archive23basic_binary_oprimitiveINS0_15binary_oarchiveEcSt11char_traitsIcEE4initEv'
/home/doried/tran/boost_build/lib/libboost_serialization.so.1.61.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status

    "g++" -L"/home/doried/tran/boost_build/lib" -L"/home/doried/tran/boost_build/lib64" -L"/home/doried/tran/oxlm/lib" -L"/home/doried/tran/oxlm/lib64" -Wl,-R -Wl,"/home/doried/tran/boost_build/lib" -Wl,-R -Wl,"/home/doried/tran/boost_build/lib64" -Wl,-R -Wl,"/home/doried/tran/mosesdecoder/moses/bin/gcc-4.8/release/threading-multi" -Wl,-R -Wl,"/home/doried/tran/oxlm/lib" -Wl,-R -Wl,"/home/doried/tran/oxlm/lib64" -Wl,-rpath-link -Wl,"/home/doried/tran/boost_build/lib" -Wl,-rpath-link -Wl,"/home/doried/tran/boost_build/lib64" -Wl,-rpath-link -Wl,"/home/doried/tran/mosesdecoder/moses/bin/gcc-4.8/release/threading-multi" -Wl,-rpath-link -Wl,"/home/doried/tran/oxlm/lib" -Wl,-rpath-link -Wl,"/home/doried/tran/oxlm/lib64" -o "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/BackwardTest" -Wl,--start-group "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/BackwardTest.o" "util/double-conversion/bin/gcc-4.8/release/threading-multi/bignum-dtoa.o" "util/double-conversion/bin/gcc-4.8/release/threading-multi/fast-dtoa.o" "util/double-conversion/bin/gcc-4.8/release/threading-multi/diy-fp.o" "util/double-conversion/bin/gcc-4.8/release/threading-multi/double-conversion.o" "util/double-conversion/bin/gcc-4.8/release/threading-multi/bignum.o" "util/bin/gcc-4.8/release/threading-multi/file.o" "util/bin/gcc-4.8/release/threading-multi/pool.o" "util/bin/gcc-4.8/release/threading-multi/integer_to_string.o" "util/bin/gcc-4.8/release/threading-multi/file_piece.o" "util/bin/gcc-4.8/release/threading-multi/ersatz_progress.o" "util/bin/gcc-4.8/release/threading-multi/mmap.o" "util/bin/gcc-4.8/release/threading-multi/exception.o" "util/bin/gcc-4.8/release/threading-multi/usage.o" "util/bin/gcc-4.8/release/threading-multi/float_to_string.o" "util/bin/gcc-4.8/release/threading-multi/bit_packing.o" "util/bin/gcc-4.8/release/threading-multi/random.o" "util/bin/gcc-4.8/release/threading-multi/murmur_hash.o" "util/bin/gcc-4.8/release/threading-multi/string_piece.o" "util/bin/gcc-4.8/release/threading-multi/scoped.o" "lm/bin/gcc-4.8/release/threading-multi/virtual_interface.o" "lm/bin/gcc-4.8/release/threading-multi/read_arpa.o" "lm/bin/gcc-4.8/release/threading-multi/bhiksha.o" "lm/bin/gcc-4.8/release/threading-multi/model.o" "lm/bin/gcc-4.8/release/threading-multi/value_build.o" "lm/bin/gcc-4.8/release/threading-multi/binary_format.o" "lm/bin/gcc-4.8/release/threading-multi/trie.o" "lm/bin/gcc-4.8/release/threading-multi/quantize.o" "lm/bin/gcc-4.8/release/threading-multi/vocab.o" "lm/bin/gcc-4.8/release/threading-multi/config.o" "lm/bin/gcc-4.8/release/threading-multi/search_hashed.o" "lm/bin/gcc-4.8/release/threading-multi/sizes.o" "lm/bin/gcc-4.8/release/threading-multi/trie_sort.o" "lm/bin/gcc-4.8/release/threading-multi/search_trie.o" "lm/bin/gcc-4.8/release/threading-multi/lm_exception.o" "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/Backward.o" "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/BackwardLMState.o" "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/Base.o" "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/BilingualLM.o" "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/Implementation.o" "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/Ken.o" "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/MultiFactor.o" "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/Remote.o" "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/SingleFactor.o" "moses/LM/bin/BackwardTest.test/gcc-4.8/release/threading-multi/SkeletonLM.o" "moses/LM/bin/gcc-4.8/release/threading-multi/OxLM.o" "moses/LM/bin/gcc-4.8/release/threading-multi/SourceOxLM.o" "moses/LM/bin/gcc-4.8/release/threading-multi/OxLMMapper.o" "moses/LM/bin/gcc-4.8/release/threading-multi/OxLMParallelMapper.o" "util/bin/gcc-4.8/release/threading-multi/read_compressed.o" "util/bin/gcc-4.8/release/threading-multi/parallel_read.o" "util/double-conversion/bin/gcc-4.8/release/threading-multi/strtod.o" "util/double-conversion/bin/gcc-4.8/release/threading-multi/cached-powers.o" "util/double-conversion/bin/gcc-4.8/release/threading-multi/fixed-dtoa.o" "moses/bin/gcc-4.8/release/threading-multi/libmoses.so"  -Wl,-Bstatic  -Wl,-Bdynamic -lboost_thread -lboost_system -lboost_unit_test_framework -lmurmurhash -llbl -lboost_filesystem -lbz2 -lgomp -lboost_program_options -lSegFault -lz -lboost_thread -lboost_system -lrt -Wl,--end-group -pthread 

BOOST 1.61.0

**Comment by doried-a-a** [comment]
Hi!

I've solved my issue with the following:

- Firstly, bjam builder was issuing some g++ build commands without passing the **libboost_serialization** to them. So I fixed it with a small change on _Jamroot_ (removed the line mentioning boost_serialozation as an external library).

- There was another problem, prevented moses from building with OxLM, which is in the file
`moses/TranslationModel/UG/mm/ug_typedefs.h `: here they defined a type named **`iptr`** ,
which was conflicting with a variable name in Eigen library used by OXLM.

I've fixed the problems in moses repository, and I've sent them a pull request, hoping they will accept it, so the problem be solved.

Best wishes!
